packages/playwright-core/src/server/network.ts

```diff
@@ -286,12 +292,17 @@ export class Route extends SdkObject {
     const headers = [...(overrides.headers || [])];
     this._maybeAddCorsHeaders(headers);
     this._request._context.emit(BrowserContext.Events.RequestFulfilled, this._request);
-    await this._delegate.fulfill({
-      status: overrides.status || 200,
-      headers,
-      body,
-      isBase64,
-    });
+    await Promise.race([
+      this._delegate.fulfill({
+        status: overrides.status || 200,
+        headers,
+        body,
+        isBase64,
+      }),
+      // If the request is already cancelled by the page before we handle the route,
+      // we'll receive loading failed event and will ignore route handling error.
+      this._request.response()
```

Comment:
Let's have a method `_waitForRequestFailed` that would explicitly wait for the request to fail?